Kernel/Net/NE2000NetworkAdapter.cpp

```diff
@@ -0,0 +1,487 @@
+    out8(REG_WR_INTERRUPTMASK, interrupt_mask);
+    rdma_read(0, Bytes(prom, sizeof(prom)));
+    for (int i = 0; i < 6; i++) {
+        m_mac_address[i] = prom[i * 2];
+    }
+
+    out8(REG_RW_COMMAND, BIT_COMMAND_PAGE1 | BIT_COMMAND_DMA_ABORT | BIT_COMMAND_STOP);
+    for (int i = 0; i < 6; i++) {
+        out8(REG_RW_PHYSICALADDRESS0 + i, m_mac_address[i]);
+    }
+    out8(REG_RW_CURRENT, NE2K_RAM_RECV_BEGIN >> 8);
+
+    out8(REG_RW_COMMAND, BIT_COMMAND_DMA_ABORT | BIT_COMMAND_START);
+    out8(REG_WR_TRANSMITCONFIGURATION, 0xe0);
+}
+
+void NE2000NetworkAdapter::rdma_read(size_t address, Bytes payload)
+{
+#if NE2000_DEBUG
+    klog() << "NE2000NetworkAdapter: DMA read @ " << PhysicalAddress(address) << " length=" << payload.size();
+#endif
+    u8 command = in8(REG_RW_COMMAND) & ~(BIT_COMMAND_PAGE_FIELD | BIT_COMMAND_DMA_FIELD);
+    out8(REG_RW_COMMAND, command | BIT_COMMAND_DMA_ABORT);
+    out8(REG_RW_INTERRUPTSTATUS, BIT_INTERRUPTMASK_RDC);
+
+    out8(REG_WR_REMOTEBYTECOUNT0, payload.size());
+    out8(REG_WR_REMOTEBYTECOUNT1, payload.size() >> 8);
+    out8(REG_WR_REMOTESTARTADDRESS0, address);
+    out8(REG_WR_REMOTESTARTADDRESS1, address >> 8);
+
+    command = in8(REG_RW_COMMAND) & ~(BIT_COMMAND_DMA_FIELD);
+    out8(REG_RW_COMMAND, command | BIT_COMMAND_DMA_READ);
+
+    for (size_t i = 0; i < payload.size(); i += 2) {
+        u16 data = in16(REG_RW_IOPORT);
+        payload[i] = data;
+        if (i != payload.size() - 1)
+            payload[i + 1] = data >> 8;
+    }
+
+    while (!(in8(REG_RW_INTERRUPTSTATUS) & BIT_INTERRUPTMASK_RDC))
+        ;
+}
+
+void NE2000NetworkAdapter::rdma_write(size_t address, ReadonlyBytes payload)
+{
+#if NE2000_DEBUG
+    klog() << "NE2000NetworkAdapter: DMA write @ " << PhysicalAddress(address) << " length=" << payload.size();
+#endif
+    u8 command = in8(REG_RW_COMMAND) & ~(BIT_COMMAND_PAGE_FIELD | BIT_COMMAND_DMA_FIELD);
+    out8(REG_RW_COMMAND, command | BIT_COMMAND_DMA_ABORT);
+    out8(REG_RW_INTERRUPTSTATUS, BIT_INTERRUPTMASK_RDC);
+
+    out8(REG_WR_REMOTEBYTECOUNT0, payload.size());
+    out8(REG_WR_REMOTEBYTECOUNT1, payload.size() >> 8);
+    out8(REG_WR_REMOTESTARTADDRESS0, address);
+    out8(REG_WR_REMOTESTARTADDRESS1, address >> 8);
+
+    command = in8(REG_RW_COMMAND) & ~(BIT_COMMAND_DMA_FIELD);
+    out8(REG_RW_COMMAND, command | BIT_COMMAND_DMA_WRITE);
+
+    for (size_t i = 0; i < payload.size(); i += 2) {
+        u16 data = payload[i];
+        if (i != payload.size() - 1)
+            data |= payload[i + 1] << 8;
+        out16(REG_RW_IOPORT, data);
+    }
+
+    while (!(in8(REG_RW_INTERRUPTSTATUS) & BIT_INTERRUPTMASK_RDC))
+        ;
+}
+
+void NE2000NetworkAdapter::send_raw(ReadonlyBytes payload)
+{
+#if NE2000_DEBUG
+    klog() << "NE2000NetworkAdapter: Sending packet length=" << payload.size();
+#endif
+
+    if (payload.size() > NE2K_RAM_SEND_SIZE) {
+        klog() << "NE2000NetworkAdapter: Packet to send was too big; discarding";
```

Comment:
Please do an early return here so we don't have to wrap most of the function in `else`.